### ISSUE TITLE
Hangs after TUI background detection

### ISSUE THREAD

**Issue by tacho** [neovim/neovim#9675]
- `nvim --version`: NVIM v0.4.0-dev at 2816bc8620ca373145842641ce77805f4321b2bc
- Vim (version: ) N/A
- Operating system/version: macOS 10.13 and 10.14
- Terminal name/version: iTerm 3.3.20190304-nightly, Kitty 0.13.3, Terminal.app
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC

```

N/A, see below.

### Actual behaviour

Hello,

after commit 246408621c405a6c0f24e2e5cd66e8776eb04fe6 (Merge #9509 'TUI: detect background color') Neovim on macOS started hanging for upwards of a minute when started with a file from a Git repository as an argument.

After disabling plugins, I narrowed it down to GitGutter. However, an issue there (https://github.com/airblade/vim-gitgutter/issues/589) seems to suggest that the issue isn't in GitGutter per se. I bisected Neovim commits, and 246408621c405a6c0f24e2e5cd66e8776eb04fe6 and onwards is broken, everything before that - works fine.

If Neovim (even commits after 246408621c405a6c0f24e2e5cd66e8776eb04fe6) is started with no arguments - and then a Git-tracked file is opened from within - everything, including GitGutter, works fine. It's only the interaction between the terminal background query stalling for some reason, and GitGutter trying to execute an async job at the same time that results in the pseudo hang. If one waits it out, everything works, but it is very inconvenient to do so.

The issue only happens - for me at least - using iTerm and Kitty on macOS. Using Kitty and Konsole on Linux works fine.

Let me know if any other testing is required. Thanks.

### Expected behaviour

Neovim and GitGutter working fine.

**Comment by stinovlas** [comment]
I have the same problem using Ubuntu 18.04 and xfce4-terminal.

**Comment by dm1try** [comment]
it looks like this could be reproducible without `GitGutter`
`nvim -c "sleep 5m"`
the related [GitGutter code](https://github.com/airblade/vim-gitgutter/blob/1d422b9f98194e38bc56e54192c9bc66d95c21f1/autoload/gitgutter/diff.vim#L70-L73)

**Comment by justinmk** [comment]
@tacho Thanks for bisecting. I would guess the problem is this logic:

https://github.com/neovim/neovim/commit/d3dc94615523a71606edcbe6bf528f8b49a78c7d#diff-9fe05c714b192fe85bb903baf3355196R361

Can someone try removing that `if (starting) { ... }` block to see if it fixes the problem?

**Comment by tacho** [comment]
@justinmk thanks for the fast response! Yes, after removing that `if (starting)` block - the hang disappears.

**Comment by justinmk** [comment]
Actually, how about this patch:

```
diff --git a/src/nvim/tui/input.c b/src/nvim/tui/input.c
index 3eb88366d610..9ed3413d0af6 100644
--- a/src/nvim/tui/input.c
+++ b/src/nvim/tui/input.c
@@ -359,7 +359,7 @@ static void set_bg_deferred(void **argv)
   char *bgvalue = argv[0];
   if (starting) {
     // Wait until after startup, so OptionSet is triggered.
-    loop_schedule(&main_loop, event_create(set_bg_deferred, 1, bgvalue));
+    loop_schedule_deferred(&main_loop, event_create(set_bg_deferred, 1, bgvalue));
     return;
   }
   if (!option_was_set("bg") && !strequal((char *)p_bg, bgvalue)) {
```

**Comment by tacho** [comment]
With that one (`loop_schedule_deferred`) the hang is still there but is *much* shorter, around a second. Still noticeable, though.

**Comment by tacho** [comment]
Actually, after some more testing - when using the `loop_schedule_deferred` patch, the hang can be anything between a second and 20 seconds. Not sure what the duration is depending on though.

The original code is around a minute at least.

**Comment by justinmk** [comment]
Thanks. Yeah, that's an interesting result but not a real solution. I have an idea how to fix it.

**Comment by aktau** [comment]
I think I'm seeing this too, on Debian Testing. Syncing to the latest merge before d3dc946 stops the hanging so that seems to be the issue indeed.

**Comment by blueyed** [comment]
Just for reference, I've left more info at https://github.com/neovim/neovim/pull/9509#issuecomment-470053342 with regard to vim-gitgutter.

But I can reproduce it also already using just `nvim -u NORC -c "sleep 5m"` - isn't that reproducible for you then, @justinmk ?
Both of the suggested patches do not help in this regard.

btw: it should probably also not cause a busy loop (using all CPU) in the first place.

The following patch fixes it for me:
```diff
diff --git i/src/nvim/tui/input.c w/src/nvim/tui/input.c
index 3eb88366d..939672863 100644
--- i/src/nvim/tui/input.c
+++ w/src/nvim/tui/input.c
@@ -11,6 +11,7 @@
 #include "nvim/main.h"
 #include "nvim/aucmd.h"
 #include "nvim/option.h"
+#include "nvim/ex_docmd.h"
 #include "nvim/os/os.h"
 #include "nvim/os/input.h"
 #include "nvim/event/rstream.h"
@@ -359,7 +360,8 @@ static void set_bg_deferred(void **argv)
   char *bgvalue = argv[0];
   if (starting) {
     // Wait until after startup, so OptionSet is triggered.
-    loop_schedule(&main_loop, event_create(set_bg_deferred, 1, bgvalue));
+    do_sleep(100);
+    loop_schedule_deferred(&main_loop, event_create(set_bg_deferred, 1, bgvalue));
     return;
   }
   if (!option_was_set("bg") && !strequal((char *)p_bg, bgvalue)) {
```
(If `do_sleep` gets used here in the end, it should probably be moved for saner includes)

**Comment by dm1try** [comment]
@blueyed `do_sleep` does not look like a solution as it will affect startup time.

this change might help:
```diff
diff --git a/src/nvim/tui/input.c b/src/nvim/tui/input.c
index 3eb88366d..ecd3180bf 100644
--- a/src/nvim/tui/input.c
+++ b/src/nvim/tui/input.c
@@ -358,8 +358,9 @@ static void set_bg_deferred(void **argv)
 {
   char *bgvalue = argv[0];
   if (starting) {
+    loop_poll_events(&main_loop, 1);
     // Wait until after startup, so OptionSet is triggered.
-    loop_schedule(&main_loop, event_create(set_bg_deferred, 1, bgvalue));
+    loop_schedule_deferred(&main_loop, event_create(set_bg_deferred, 1, bgvalue));
     return;
   }
   if (!option_was_set("bg") && !strequal((char *)p_bg, bgvalue)) {
```
though it is also dirty as `set_bg_deffered` will be re-scheduled many times for running during "starting" phase. 

it looks like some synchronization problem: we would like to do some work after startup with the data that we received during startup(the background color received during startup should be set after startup to guarantee the triggering of `OptionSet`).
maybe the loop API can be extended?

**Comment by przepompownia** [comment]
> reproducible without `GitGutter`: `nvim -u NORC -c "sleep 5m"` (only with kitty and iTerm)
> the related [GitGutter code](https://github.com/airblade/vim-gitgutter/blob/1d422b9f98194e38bc56e54192c9bc66d95c21f1/autoload/gitgutter/diff.vim#L70-L73)

Xterm, rxvt-unicode and gnome-terminal are also affected by this bug, but tmux and stterm - not. The bug does not occur in xterm after switching `XTerm*allowSendEvents` to true, but this option is described as insecure and in addition mouse does not work.

**Comment by ilAYAli** [comment]
@dm1try your suggested changes seems to solve the problem for me.

**Comment by tex** [comment]
I tested @dm1try patch on Kitty / NixOS and it seems to fix the problem for me.

**Comment by justinmk** [comment]
see #9708

**Comment by tacho** [comment]
@justinmk - I just compiled with the patch in #9708 applied, and while it no longer hangs, it also no longer correctly reports the background (light/dark). With unpatched Neovim, I get the hang, but after it eventually starts up - background is set correctly to light or dark, depending on my iTerm colorscheme.

With this patch, startup delay is unnoticeable, if any, but background is always set to dark, regardless of the terminal theme. Verified without user init files as well as with.
